Marlin/src/gcode/calibrate/G28.cpp

```diff
@@ -319,7 +319,7 @@ void GcodeSuite::G28() {
     if (z_homing_height && (doX || doY || (ENABLED(Z_SAFE_HOMING) && doZ))) {
       // Raise Z before homing any other axes and z is not already high enough (never lower z)
       if (DEBUGGING(LEVELING)) DEBUG_ECHOLNPAIR("Raise Z (before homing) by ", z_homing_height);
-      do_z_clearance(z_homing_height, TEST(axis_known_position, Z_AXIS), DISABLED(UNKNOWN_Z_NO_RAISE));
+      do_z_clearance(z_homing_height, (TEST(axis_known_position, Z_AXIS) || (current_position.z >= z_homing_height && Z_ENABLE_READ() == Z_ENABLE_ON)), DISABLED(UNKNOWN_Z_NO_RAISE));
```

Comment:
The `Z_ENABLE_READ() == Z_ENABLE_ON` state isn't trustworthy. A stepper motor might get powered back on after it has been powered off, and the Z enable pin state won't tell us about that.

So `TEST(axis_known_position, Z_AXIS)` should be used instead of `Z_ENABLE_READ() == Z_ENABLE_ON` here.

Reply:
`TEST(axis_known_position, Z_AXIS) || (current_position.z >= z_homing_height && TEST(axis_known_position, Z_AXIS))`
and
`TEST(axis_known_position, Z_AXIS)`
is the same thing so no need to change
once z pos is know move is always absolute. Only solution I can find is to store in a bool if raise as been already done and clear it when Z is disabled

Reply:
> Only solution I can find is to store in a bool if raise as been already done and clear it when Z is disabled

OK, in my opinion this would be the proper solution.
I am not that expertised in CPP and can not do it.

Reply:
@swissnorp please note that my comment was about fix proposed by @thinkyhead. I understand why you checked motor 'enabled', because after first raise this is its state and I also understand what is the @thinkyhead thinking. He is right about "knowing" state but this will lead to a continuos raise (until Z is properly homed) everytime X or Y is to be homed.
Both solutions are...not a solution. In your proposed solution, if stepper is at 10mm, then it's disengag and falls to the bed and someone does a jog (0,01mm?) postion is 10.001, head is at 0.001 but motor is enabled result: You don't rise but you should.
Since after de-energization there is noone that can tell you where head is really positionated, you have to rise (even if cur pos is big), once you have done the rise, then no more rises until axis will not be de-energized. In my opinion this may be done only with a flag that store a "rise done after last enable" that must be reset when stepper is disabled.

Reply:
On certain machines, before Z has been homed its position is unknown, but it is known that Z is at its lowest or perhaps its highest point, but almost certainly above the bed. The raise before homing was I believe intended to be to an absolute position and not just relative every time Z is "not known." I think it is in a sense safe for these raises during first-time homing to say Z is "sort of known" or at least "clearing the clips."

I feel like it might have been expressly this way in a past version of `G28` but I may be mistaken.

I can't see any obvious drawback to treating the `Z_HOMING_HEIGHT` as always being a Z position above 0, and then no extra flag wis needed. (The current Z position is the flag.)

Reply:
This seems to do the trick, but is it too simple to be true? adf87fbc6aa1269ce9ac2b010f187f143c3434bc

Reply:
I don't like this solution because it works with "absolute" position. If Z is at 100 and I home X it may go to 5...I may panic seeing Z moving down